bindings/rust/src/lib.rs

```diff
@@ -665,6 +669,29 @@ macro_rules! sig_variant_impl {
             }
         }
 
+        #[cfg(feature = "serde")]
+        impl Serialize for SecretKey {
+            fn serialize<S: Serializer>(
+                &self,
+                ser: S,
+            ) -> Result<S::Ok, S::Error> {
+                let bytes = Zeroizing::new(self.serialize());
+                ser.serialize_bytes(bytes.as_ref())
+            }
+        }
+
+        #[cfg(feature = "serde")]
+        impl<'de> Deserialize<'de> for SecretKey {
+            fn deserialize<D: Deserializer<'de>>(
+                deser: D,
+            ) -> Result<Self, D::Error> {
+                let bytes: &[u8] = Deserialize::deserialize(deser)?;
+                Self::deserialize(&bytes).map_err(|e| {
```

Comment:
It's as appropriate to zeroize `bytes` here as in serialization. Though I have no suggestions for how. (BTW, one can pass `bytes` here instead of `&bytes`.)

Reply:
Likewise, I don't see a good way to zeroize the buffer. By asking for `&[u8]` we might even get zero-copy access to the underlying buffer, which the `Deserializer` instance doesn't have mutable access to.

It might be that the right approach here is to feature gate this with `serde_secret` and put a big warning about being careful to zeroize serialized secret keys. Or just not implementing ser/des for secrets, I suppose---though I worry that people will do it anyway and potentially make things even worse...

Reply:
Added a separate feature gate for `serde-secret`. I'm not 100% sure this is the right approach---maybe instead there should be no serde support for SecretKey. Sadly, I still don't see how to forcibly zeroize the bytes being deserialized.

Reply:
> I still don't see how to forcibly zeroize the bytes being deserialized.

It might be by design. I mean, as @nazar-pc hinted, if zero-copy is a possibility, then the interface is bound to provide non-mutable access only. As a way to look at it. One can coerce the mutability with `unsafe`, but if the underlying deserializer happens to be zero-copy, then you'd be doing real damage.

Reply:
For zero copy to work the data structure itself must have references to original data (also `Copy` types are not actually considered copying for some reason). If this data structure doesn't have reference to original data, then deserializing slice of memory doesn't make that much sense, just deserialize fixed array or whatever you actually need here.

Reply:
Just in case, in deserializion context zero-copy would mean that you would get a direct reference to the original data, right? In this case 32 bytes suitable for deserialization into SecretKey. It would be inappropriate to zero the original, right? Or are you suggesting that one can make serde write data into the SecretKey and then execute Self::deserialize in place? The C side would handle it. So if this is possible, do enlighten us.

Moreover, one can even question why would one call Self::deserialize and Self::serialize at all? Just let serde handle a sequence of 32 bytes as it is. It would work, because the blst_scalar's internal representation is portable among platforms.

Reply:
> Just in case, in deserializion context zero-copy would mean that you would get a direct reference to the original data, right? In this case 32 bytes suitable for deserialization into SecretKey.

Either that or statically known `Copy` value (in Rust at least). In other words, no allocation on the heap is necessary.

> It would be inappropriate to zero the original, right?

Exactly, deserialized value doesn't own data in that case, it just points to existing memory. Inn this case though if it is just statically known array of 32 bytes, there isn't much point to talk about zero copy, you just read from source and create `SecretKey` instance. It is up to the provider of the input to erase input correctly afterwards IMO.

> Just let serde handle a sequence of 32 bytes as it is.

Yep, that is what I'd do. Deserialize 32 bytes and create `SecretKey` out of it. I don't really see why it is necessary to create unsized slice of memory first, check its length again and then create an instance by copying data one more time.

Reply:
The reason I was calling `Self::deserialize` and `Self::serialize` is that I wanted to preserve any invariants under serialization/deserialization that these functions enforce.

I'm guessing `Self::serialize` doesn't do any checks, so probably we can get rid of that. But doesn't `Self::deserialize` check that the value passed in is less than the order of the curve? If so, we should enforce that upon deserialization, no?

Reply:
> If so, we should enforce that upon deserialization, no?

I believe it is necessary to enforce indeed. I didn't look into implementation though to suggest something specific here.

Reply:
> But doesn't Self::deserialize check that the value passed in is less than the order of the curve?

Yeah, it does... Well, it's a function call. So that if there is a way to bypass the intermediate non-zeroize-able buffer as the cost of making the call in question, it would be a win...